modules/apps/static/portal/portal-db-partition-impl/src/main/java/com/liferay/portal/db/partition/internal/DBPartitionHelperImpl.java

```diff
@@ -0,0 +1,212 @@
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.dao.db.DB;
+import com.liferay.portal.kernel.dao.db.DBInspector;
+import com.liferay.portal.kernel.dao.db.DBManagerUtil;
+import com.liferay.portal.kernel.dao.db.DBType;
+import com.liferay.portal.kernel.dao.db.partition.DBPartitionHelper;
+import com.liferay.portal.kernel.dao.jdbc.CurrentConnectionUtil;
+import com.liferay.portal.kernel.dao.orm.ORMException;
+import com.liferay.portal.kernel.model.CompanyConstants;
+import com.liferay.portal.kernel.security.auth.CompanyThreadLocal;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.InfrastructureUtil;
+import com.liferay.portal.kernel.util.PropsUtil;
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Alberto Chaparro
+ */
+@Component(immediate = true, service = {})
+public class DBPartitionHelperImpl implements DBPartitionHelper {
+
+	@Override
+	public boolean addPartition(long companyId) {
+		if (companyId == _defaultCompanyId) {
+			return false;
+		}
+
+		Connection connection = CurrentConnectionUtil.getConnection(
+			InfrastructureUtil.getDataSource());
+
+		String schemaName = "company" + companyId;
+
+		try {
+			try (PreparedStatement preparedStatement =
+					connection.prepareStatement(
+						"create schema if not exists " + schemaName +
+							" character set utf8")) {
+
+				preparedStatement.executeUpdate();
+			}
+
+			DatabaseMetaData databaseMetaData = connection.getMetaData();
+
+			DBInspector dbInspector = new DBInspector(connection);
+
+			try (ResultSet resultSet = databaseMetaData.getTables(
+					dbInspector.getCatalog(), dbInspector.getSchema(), null,
+					new String[] {"TABLE"});
+				Statement statement = connection.createStatement()) {
+
+				while (resultSet.next()) {
+					String tableName = resultSet.getString("TABLE_NAME");
+
+					if (_isControlTable(dbInspector, tableName)) {
+						statement.executeUpdate(
+							StringBundler.concat(
+								"create view ", schemaName, StringPool.PERIOD,
+								tableName, " as select * from companyDefault.",
+								tableName));
+					}
+					else {
+						statement.executeUpdate(
+							StringBundler.concat(
+								"create table ", schemaName, StringPool.PERIOD,
+								tableName, " like companyDefault.", tableName));
+					}
+				}
+			}
+		}
+		catch (Exception exception) {
+			throw new ORMException(exception);
+		}
+
+		return true;
+	}
+
+	@Override
+	public boolean removePartition(long companyId) {
+		return true;
```

Comment:
@achaparro is this some incompleted code?

Reply:
Yes and no, Mike wants to handle the company removal in a separate JVM. This is a temporary solution to avoid failing when we try to remove a schema with db partitioning. We will have to implement a separate process to remove companies mark previously as inactive.